integration/container_stats_test.go

```diff
@@ -0,0 +1,317 @@
+		defer func() {
+			assert.NoError(t, runtimeService.StopContainer(cn, 10))
+		}()
+	}
+
+	t.Logf("Fetch container stats for each containerd with Filter")
+	var stats []*runtime.ContainerStats
+	for i := 0; i < 3; i++ {
+		require.NoError(t, Eventually(func() (bool, error) {
+			stats, err = runtimeService.ListContainerStats(
+				&runtime.ContainerStatsFilter{Id: containers[i]})
+			if err != nil {
+				return false, err
+			}
+			if len(stats) != 1 {
+				return false, fmt.Errorf("unexpected stats length")
+			}
+			if stats[0].GetWritableLayer().GetUsedBytes().GetValue() != 0 &&
+				stats[0].GetWritableLayer().GetInodesUsed().GetValue() != 0 {
+				return true, nil
+			}
+			return false, nil
+		}, time.Second, 30*time.Second))
+
+		t.Logf("Verify container stats for %s", containers[i])
+		for _, s := range stats {
+			require.Equal(t, s.GetAttributes().GetId(), containers[i])
+			testStats(t, s)
+		}
+	}
+}
+
+// Test to verify filtering given a specific Sandbox ID. Stats for
+// all the containers in a pod should be returned
+func TestContainerListStatsWithSandboxIdFilter(t *testing.T) {
+	t.Logf("Create a pod config and run sandbox container")
+	sbConfig := PodSandboxConfig("running-pod", "statsls")
+	sb, err := runtimeService.RunPodSandbox(sbConfig)
+	require.NoError(t, err)
+	defer func() {
+		assert.NoError(t, runtimeService.StopPodSandbox(sb))
+		assert.NoError(t, runtimeService.RemovePodSandbox(sb))
+	}()
+	t.Logf("Create a container config and run containers in a pod")
+	containers := []string{}
+	opt := func() ContainerOpts {
+		return func(cf *runtime.ContainerConfig) {
+			cf.Labels = map[string]string{"key": "value"}
+			cf.Annotations = map[string]string{"a.b.c": "test"}
+		}
+	}()
+	for i := 0; i < 3; i++ {
+		cName := fmt.Sprintf("container%d", i)
+		containerConfig := ContainerConfig(cName, defaultTestImage, opt)
+		cn, err := runtimeService.CreateContainer(sb, containerConfig, sbConfig)
+		containers = append(containers, cn)
+		require.NoError(t, err)
+		defer func() {
+			assert.NoError(t, runtimeService.RemoveContainer(cn))
+		}()
+		require.NoError(t, runtimeService.StartContainer(cn))
+		defer func() {
+			assert.NoError(t, runtimeService.StopContainer(cn, 10))
+		}()
+	}
+
+	t.Logf("Fetch container stats for each containerd with Filter")
+	var stats []*runtime.ContainerStats
+	require.NoError(t, Eventually(func() (bool, error) {
+		stats, err = runtimeService.ListContainerStats(
+			&runtime.ContainerStatsFilter{PodSandboxId: sb})
+		if err != nil {
+			return false, err
+		}
+		if len(stats) != 3 {
+			return false, fmt.Errorf("unexpected stats length")
+		}
+		if stats[0].GetWritableLayer().GetUsedBytes().GetValue() != 0 &&
+			stats[0].GetWritableLayer().GetInodesUsed().GetValue() != 0 {
+			return true, nil
+		}
+		return false, nil
+	}, time.Second, 30*time.Second))
+	t.Logf("Verify container stats for sandbox %q", sb)
+	for _, s := range stats {
+		testStats(t, s)
+	}
+}
+
```

Comment:
Thought process here is that its fine not to be conservative with tests especially with this scenario not being exercised by e2e tests. Unit test is good to test the filtering I agree. I would like to add a TODO here to take this off once we have e2e coverage for these functionalities. Always happy to remove code :) Let me know.

Reply:
I feel like that we should:
1) Add comprehensive unit test for filtering even no matter we test filter or not here.
2) There are many duplicated code in this file, we could easily use a table-driven test pattern to get rid of these duplication.

However, I don't want to block this PR because of test code, so I'm fine with adding TODO for now, and come back later.